cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+{
+    if (package_info)
+    {
+        free(package_info->arch);
+        free(package_info->name);
+        free(package_info->version);
+
+        free(package_info);
+    }
+}
+
+static char *ParseOptions(Rlist *options)
+{
+    if (RlistIsNullList(options))
+    {
+        return SafeStringDuplicate("");
+    }
+    
+    Buffer *data = BufferNew();
+    for (Rlist *rp = options; rp != NULL; rp = rp->next)
+    {
+        char *value = RlistScalarValue(rp);
+        BufferAppendString(data, "options=");
+        BufferAppendString(data, value);
+        BufferAppendString(data, "\n");
+    }
+    return BufferClose(data);
+}
+
+static PackageInfo *ParseAndCheckPackageDataReply(const Rlist *data)
+{
+    PackageInfo * package_data = xcalloc(1, sizeof(PackageInfo));
+    
+    for (const Rlist *rp = data; rp != NULL; rp = rp->next)
+    {
+        char *line = RlistScalarValue(rp);
+                   
+        if (StringStartsWith(line, "PackageType="))
+        {
+            char *type = line + strlen("PackageType=");
+            if (StringSafeEqual(type, "file"))
+            {
+                package_data->type = PACKAGE_TYPE_FILE;
+            }
+            else if (StringSafeEqual(type, "repo"))
+            {
+                package_data->type = PACKAGE_TYPE_REPO;
+            }
+            else
+            {
+                Log(LOG_LEVEL_VERBOSE, "unsupported package type: %s", type);
+                free(package_data);
+                return NULL;
+            }
+        }
+        else if (StringStartsWith(line, "Name="))
+        {
+            if (package_data->name)
+            {
+                /* Some error occurred as we already have name for 
+                 * given package. */
+                Log(LOG_LEVEL_ERR, "Duplicated package name "
+                    "received: [%s] %s", line, package_data->name);
+                free(package_data);
+                return NULL;
+            }
+            package_data->name = 
+                SafeStringDuplicate(line + strlen("Name="));
+        }
+        else if (StringStartsWith(line, "Version="))
+        {
+            if (package_data->version)
+            {
+                /* Some error occurred as we already have version for 
+                 * given package. */
+                Log(LOG_LEVEL_ERR, "Duplicated package version "
+                    "received: [%s] %s", line, package_data->version);
+                free(package_data);
+                return NULL;
+            }
+            package_data->version = 
+                SafeStringDuplicate(line + strlen("Version="));
+        }
+        else if (StringStartsWith(line, "Architecture="))
+        {
+            if (package_data->arch)
+            {
+                /* Some error occurred as we already have arch for 
+                 * given package. */
+               Log(LOG_LEVEL_ERR, "Duplicated package architecture "
+                    "received: [%s] %s", line, package_data->arch);
```

Comment:
Duplicated is indicating that we have received second line with the same parameter description. We are expecting ONLY one so "Duplicated name description received" means that we already had one line describing name of the package. So it is not indicating that name is exactly the same but line containing package name description is seen 2nd time.